vortex-array/src/stats/bound.rs

```diff
@@ -0,0 +1,322 @@
+use std::cmp::Ordering;
+
+use vortex_error::{VortexError, VortexResult};
+use Precision::Inexact;
+
+use crate::stats::Precision::Exact;
+use crate::stats::{Precision, StatBound};
+
+/// Interpret the value as a lower bound.
+/// These form a partial order over successively more precise bounds
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct LowerBound<T>(pub(crate) Precision<T>);
+
+impl<T> LowerBound<T> {
+    pub(crate) fn min_value(self) -> T {
+        self.0.into_value()
+    }
+}
+
+impl<T> LowerBound<T> {
+    pub fn is_exact(&self) -> bool {
+        self.0.is_exact()
+    }
+}
+
+/// The result of the fallible intersection of two bound, defined to avoid Option JoinResult mixup.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum JoinResult<T> {
+    Join(T),
+    None,
+}
+
+impl<T> JoinResult<T> {
+    pub fn ok_or_else<F>(self, err: F) -> VortexResult<T>
+    where
+        F: FnOnce() -> VortexError,
+    {
+        match self {
+            JoinResult::Join(v) => Ok(v),
+            JoinResult::None => Err(err()),
+        }
+    }
+}
+
+impl<T: PartialOrd + Clone> StatBound<T> for LowerBound<T> {
+    fn lift(value: Precision<T>) -> Self {
+        Self(value)
+    }
+
+    // The meet or tightest covering bound
+    fn union(&self, other: &Self) -> Option<LowerBound<T>> {
+        Some(LowerBound(match (&self.0, &other.0) {
+            (Exact(lhs), Exact(rhs)) => Exact(min(lhs, rhs)?.clone()),
+            (Inexact(lhs), Inexact(rhs)) => Inexact(min(lhs, rhs)?.clone()),
+            (Inexact(lhs), Exact(rhs)) => {
+                if rhs <= lhs {
+                    Exact(rhs.clone())
+                } else {
+                    Inexact(lhs.clone())
+                }
+            }
+            (Exact(lhs), Inexact(rhs)) => {
+                if rhs >= lhs {
+                    Exact(lhs.clone())
+                } else {
+                    Inexact(rhs.clone())
+                }
+            }
+        }))
+    }
+
+    // The join of the smallest intersection of both bounds, this can fail.
+    fn intersection(&self, other: &Self) -> Option<JoinResult<LowerBound<T>>> {
```

Comment:
Ok yep, let's make this `-> VortexResult<Option<LowerBound<T>>>` and add a doc comment (you need three slashes) that says "returns an error if the bound values are not totally orderable"

Reply:
We decide to go with intersection result